### ISSUE TITLE
[parser] Error with ESLint 6.0.0 pre-releases

### ISSUE THREAD

**Issue by teppeis** [typescript-eslint/typescript-eslint#563]
**What code were you trying to parse?**

```js
module.exports = {
  overrides: [
    {
      files: ["*.ts", "*.tsx"],
      parser: require.resolve("@typescript-eslint/parser")
}
```

**What did you expect to happen?**

Able to parse TypeScript files.

**What actually happened?**

<details>
<summary>Error: Failed to load parser '@typescript-eslint/parser/dist/parser.js': Cannot find module 'eslint/lib/util/traverser'</summary>

```
Error: Failed to load parser '/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/@typescript-eslint/parser/dist/parser.js' declared in '--config » ../../+typescript.js#overrides[0]': Cannot find module 'eslint/lib/util/traverser'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:582:15)
    at Function.Module._load (internal/modules/cjs/loader.js:508:25)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at Object.<anonymous> (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/@typescript-eslint/parser/dist/parser.js:7:37)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at ConfigArrayFactory._loadParser (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:756:29)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:549:32)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:490:20)
    at _normalizeObjectConfigData.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:580:25)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:490:20)
    at _normalizeObjectConfigData.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:544:25)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:490:20)
    at _normalizeObjectConfigData.next (<anonymous>)
    at createConfigArray (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:306:25)
    at ConfigArrayFactory.loadFile (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/config-array-factory.js:381:16)
    at createCLIConfigArray (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/cascading-config-array-factory.js:140:35)
    at new CascadingConfigArrayFactory (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/cascading-config-array-factory.js:199:29)
    at new CLIEngine (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/node_modules/eslint/lib/cli-engine/cli-engine.js:552:36)
    at verify (/Users/teppei/src/github.com/teppeis/eslint-config-teppeis/test/index.js:23:18)
```
</details>

https://github.com/typescript-eslint/typescript-eslint/blob/cca1714c4bcb7a0349b4e10103870cc46bedf052/packages/parser/src/parser.ts#L8

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `1.9.0` |
| `TypeScript`                | `3.4.5` |
| `ESLint`                    | `6.0.0-alpha.2` |
| `node`                      | `10.15.3` |
| `npm`                       | `6.9.0` |

**Comment by teppeis** [comment]
Related: [indent-new-do-not-use] Error with ESLint 6.0.0-alpha.2 #564

**Comment by JamesHenry** [comment]
ESLint 6 has moved into the RC phase with rc.0

**Comment by JamesHenry** [comment]
We could probably temporarily support both ESLint 5 and 6 by converting these imports to requires and wrapping them in a try catch, what do you think @mysticatea? Are there any other changes to watch out for?

**Comment by mysticatea** [comment]
In ideal, we should not import any private files of ESLint. It's not a part of the public API, so those files can be changed anytime.

**Comment by ivan-aksamentov** [comment]
`traverser.js` has been moved from `lib/util/traverser.js` to `lib/shared/traverser.js`
in commit [eslint#219aecb](https://github.com/eslint/eslint/commit/219aecb78bc646d44bad27dc775a9b3d3dc58232) PR [eslint#11555](https://github.com/eslint/eslint/pull/11555)

As a quick fix, #614 attempts to monkey-patch this.

**Comment by JamesHenry** [comment]
FYI this should be fixed by https://github.com/typescript-eslint/typescript-eslint/pull/628
